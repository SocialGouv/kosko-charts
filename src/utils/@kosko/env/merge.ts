//

/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-explicit-any */

// HACK(douglasduteil): wrap internal @kosko/env/dist/merge.js
// As the paths API is internal (not exported) in @kosko/env
// This is code will wrap and expose it for our needs.
// see https://github.com/tommy351/kosko/blob/%40kosko/env%401.0.1/packages/env/src/merge.ts#L12-L18
// since https://github.com/tommy351/kosko/commit/6353c8559994b826c8056ad03a35332e88b9134a

// https://stackoverflow.com/a/48769843
type UnionToIntersection<T> = (T extends any ? (k: T) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;

export const merge: <T extends any[]>(
  ...data: T
) => UnionToIntersection<T[number]> =
  // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
  require("@kosko/env/dist/merge.js").merge;
