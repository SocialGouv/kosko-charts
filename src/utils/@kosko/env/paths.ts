//

import { join } from "path";

// HACK(douglasduteil): wrap internal @kosko/env/dist/paths.js
// As the paths API is internal (not exported) in @kosko/env
// This is code will wrap and expose it for our needs.
// see https://github.com/tommy351/kosko/blob/%40kosko/env%400.5.2/packages/env/src/paths.ts#L9-L13
// since https://github.com/tommy351/kosko/commit/6353c8559994b826c8056ad03a35332e88b9134a

export const formatPath: (
  path: string,
  data: Record<string, unknown>
) => string =
  // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
  require(join(require.resolve("@kosko/env"), "../paths.js")).formatPath;
