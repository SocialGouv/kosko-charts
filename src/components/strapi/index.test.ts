import { createNodeCJSEnvironment } from "@kosko/env";

import { create } from "./index";

process.env.CI_COMMIT_SHA = "longsha";
process.env.CI_COMMIT_SHORT_SHA = "shortsha";
process.env.CI_ENVIRONMENT_SLUG = "some-slug";
process.env.CI_ENVIRONMENT_NAME = "some-name";
process.env.CI_PROJECT_PATH_SLUG = "some-path";
process.env.KUBE_INGRESS_BASE_DOMAIN = "dev2.fabrique.social.gouv.fr";
process.env.KUBE_NAMESPACE = "some-namespace";

beforeEach(() => {
  jest.resetModules();
});

test("should create strapi dev config", async () => {
  const env = createNodeCJSEnvironment({ cwd: "/tmp" });
  process.env.CI_REGISTRY_IMAGE = "/path/to/docker/image";
  process.env.CI_ENVIRONMENT_URL = "https://path/to/env";
  process.env.CI_PROJECT_NAME = "some-project";
  expect(
    await create("hello", {
      env,
    })
  ).toMatchSnapshot();
});

test("should handle custom pvcName", async () => {
  const env = createNodeCJSEnvironment({ cwd: "/tmp" });
  process.env.CI_REGISTRY_IMAGE = "/path/to/docker/image";
  process.env.CI_ENVIRONMENT_URL = "https://path/to/env";
  process.env.CI_PROJECT_NAME = "some-project";
  expect(
    await create("hello", {
      config: {
        pvcName: "some-pvc",
      },
      env,
    })
  ).toMatchSnapshot();
});
