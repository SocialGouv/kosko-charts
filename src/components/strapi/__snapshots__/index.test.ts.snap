// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create strapi dev config 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "annotations": Object {
        "app.gitlab.com/app": "some-path",
        "app.gitlab.com/env": "some-slug",
        "app.gitlab.com/env.name": "some-name",
        "kapp.k14s.io/disable-default-label-scoping-rules": "",
        "kapp.k14s.io/disable-default-ownership-label-rules": "",
      },
      "labels": Object {
        "app": "hello",
        "application": "some-slug-some-project",
        "owner": "some-project",
        "team": "some-project",
      },
      "name": "hello",
      "namespace": "some-namespace",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app": "hello",
        },
      },
      "template": Object {
        "metadata": Object {
          "annotations": Object {
            "app.gitlab.com/app": "some-path",
            "app.gitlab.com/env": "some-slug",
            "app.gitlab.com/env.name": "some-name",
            "kapp.k14s.io/disable-default-label-scoping-rules": "",
            "kapp.k14s.io/disable-default-ownership-label-rules": "",
          },
          "labels": Object {
            "app": "hello",
            "application": "some-slug-some-project",
            "owner": "some-project",
            "team": "some-project",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "BACKOFFICE_URL",
                  "value": "https://strapi-some-slug-some-project.dev2.fabrique.social.gouv.fr",
                },
                Object {
                  "name": "DATABASE_CLIENT",
                  "value": "postgres",
                },
                Object {
                  "name": "DATABASE_HOST",
                  "value": "$(PGHOST)",
                },
                Object {
                  "name": "DATABASE_NAME",
                  "value": "$(PGDATABASE)",
                },
                Object {
                  "name": "DATABASE_PASSWORD",
                  "value": "$(PGPASSWORD)",
                },
                Object {
                  "name": "DATABASE_PORT",
                  "value": "$(PGPORT)",
                },
                Object {
                  "name": "DATABASE_SSL",
                  "value": "true",
                },
                Object {
                  "name": "DATABASE_USERNAME",
                  "value": "$(PGUSER)",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "azure-pg-user-sha",
                  },
                },
              ],
              "image": "/path/to/docker/image/hello:undefined",
              "livenessProbe": Object {
                "failureThreshold": 6,
                "httpGet": Object {
                  "path": "/_health",
                  "port": "http",
                },
                "initialDelaySeconds": 30,
                "periodSeconds": 5,
                "timeoutSeconds": 5,
              },
              "name": "hello",
              "ports": Array [
                Object {
                  "containerPort": 1337,
                  "name": "http",
                },
              ],
              "readinessProbe": Object {
                "failureThreshold": 15,
                "httpGet": Object {
                  "path": "/_health",
                  "port": "http",
                },
                "initialDelaySeconds": 0,
                "periodSeconds": 5,
                "successThreshold": 1,
                "timeoutSeconds": 1,
              },
              "resources": Object {
                "limits": Object {
                  "cpu": "500m",
                  "memory": "512Mi",
                },
                "requests": Object {
                  "cpu": "50m",
                  "memory": "256Mi",
                },
              },
              "startupProbe": Object {
                "failureThreshold": 12,
                "httpGet": Object {
                  "path": "/_health",
                  "port": "http",
                },
                "periodSeconds": 5,
              },
              "volumeMounts": Array [
                Object {
                  "mountPath": "/app/public/uploads",
                  "name": "uploads",
                },
              ],
            },
          ],
          "initContainers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "WAIT_FOR_RETRIES",
                  "value": "24",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "azure-pg-user-sha",
                  },
                },
              ],
              "image": "ghcr.io/socialgouv/docker/wait-for-postgres:6.0.1",
              "imagePullPolicy": "Always",
              "name": "wait-for-postgres",
              "resources": Object {
                "limits": Object {
                  "cpu": "20m",
                  "memory": "32Mi",
                },
                "requests": Object {
                  "cpu": "5m",
                  "memory": "16Mi",
                },
              },
            },
          ],
          "volumes": Array [
            Object {
              "name": "uploads",
              "persistentVolumeClaim": Object {
                "claimName": "strapi-uploads",
              },
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "annotations": Object {
        "app.gitlab.com/app": "some-path",
        "app.gitlab.com/env": "some-slug",
        "app.gitlab.com/env.name": "some-name",
        "kapp.k14s.io/disable-default-label-scoping-rules": "",
        "kapp.k14s.io/disable-default-ownership-label-rules": "",
      },
      "labels": Object {
        "app": "hello",
        "application": "some-slug-some-project",
        "owner": "some-project",
        "team": "some-project",
      },
      "name": "hello",
      "namespace": "some-namespace",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "http",
          "port": 80,
          "targetPort": 1337,
        },
      ],
      "selector": Object {
        "app": "hello",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "extensions/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "annotations": Object {
        "app.gitlab.com/app": "some-path",
        "app.gitlab.com/env": "some-slug",
        "app.gitlab.com/env.name": "some-name",
        "kapp.k14s.io/disable-default-label-scoping-rules": "",
        "kapp.k14s.io/disable-default-ownership-label-rules": "",
        "kubernetes.io/ingress.class": "nginx",
      },
      "labels": Object {
        "app": "hello",
        "application": "some-slug-some-project",
        "owner": "some-project",
        "team": "some-project",
      },
      "name": "hello",
      "namespace": "some-namespace",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "strapi-some-slug-some-project.dev2.fabrique.social.gouv.fr",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "hello",
                  "servicePort": 80,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "strapi-some-slug-some-project.dev2.fabrique.social.gouv.fr",
          ],
          "secretName": "wildcard-crt",
        },
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "PersistentVolumeClaim",
    "metadata": Object {
      "annotations": Object {},
      "name": "strapi-uploads",
    },
    "spec": Object {
      "accessModes": Array [
        "ReadWriteOnce",
      ],
      "resources": Object {
        "requests": Object {
          "storage": "5Gi",
        },
      },
      "volumeMode": "Filesystem",
    },
  },
]
`;

exports[`should handle custom pvcName 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "annotations": Object {
        "app.gitlab.com/app": "some-path",
        "app.gitlab.com/env": "some-slug",
        "app.gitlab.com/env.name": "some-name",
        "kapp.k14s.io/disable-default-label-scoping-rules": "",
        "kapp.k14s.io/disable-default-ownership-label-rules": "",
      },
      "labels": Object {
        "app": "hello",
        "application": "some-slug-some-project",
        "owner": "some-project",
        "team": "some-project",
      },
      "name": "hello",
      "namespace": "some-namespace",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app": "hello",
        },
      },
      "template": Object {
        "metadata": Object {
          "annotations": Object {
            "app.gitlab.com/app": "some-path",
            "app.gitlab.com/env": "some-slug",
            "app.gitlab.com/env.name": "some-name",
            "kapp.k14s.io/disable-default-label-scoping-rules": "",
            "kapp.k14s.io/disable-default-ownership-label-rules": "",
          },
          "labels": Object {
            "app": "hello",
            "application": "some-slug-some-project",
            "owner": "some-project",
            "team": "some-project",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "BACKOFFICE_URL",
                  "value": "https://strapi-some-slug-some-project.dev2.fabrique.social.gouv.fr",
                },
                Object {
                  "name": "DATABASE_CLIENT",
                  "value": "postgres",
                },
                Object {
                  "name": "DATABASE_HOST",
                  "value": "$(PGHOST)",
                },
                Object {
                  "name": "DATABASE_NAME",
                  "value": "$(PGDATABASE)",
                },
                Object {
                  "name": "DATABASE_PASSWORD",
                  "value": "$(PGPASSWORD)",
                },
                Object {
                  "name": "DATABASE_PORT",
                  "value": "$(PGPORT)",
                },
                Object {
                  "name": "DATABASE_SSL",
                  "value": "true",
                },
                Object {
                  "name": "DATABASE_USERNAME",
                  "value": "$(PGUSER)",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "azure-pg-user-sha",
                  },
                },
              ],
              "image": "/path/to/docker/image/hello:undefined",
              "livenessProbe": Object {
                "failureThreshold": 6,
                "httpGet": Object {
                  "path": "/_health",
                  "port": "http",
                },
                "initialDelaySeconds": 30,
                "periodSeconds": 5,
                "timeoutSeconds": 5,
              },
              "name": "hello",
              "ports": Array [
                Object {
                  "containerPort": 1337,
                  "name": "http",
                },
              ],
              "readinessProbe": Object {
                "failureThreshold": 15,
                "httpGet": Object {
                  "path": "/_health",
                  "port": "http",
                },
                "initialDelaySeconds": 0,
                "periodSeconds": 5,
                "successThreshold": 1,
                "timeoutSeconds": 1,
              },
              "resources": Object {
                "limits": Object {
                  "cpu": "500m",
                  "memory": "512Mi",
                },
                "requests": Object {
                  "cpu": "50m",
                  "memory": "256Mi",
                },
              },
              "startupProbe": Object {
                "failureThreshold": 12,
                "httpGet": Object {
                  "path": "/_health",
                  "port": "http",
                },
                "periodSeconds": 5,
              },
              "volumeMounts": Array [
                Object {
                  "mountPath": "/app/public/uploads",
                  "name": "uploads",
                },
              ],
            },
          ],
          "initContainers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "WAIT_FOR_RETRIES",
                  "value": "24",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "azure-pg-user-sha",
                  },
                },
              ],
              "image": "ghcr.io/socialgouv/docker/wait-for-postgres:6.0.1",
              "imagePullPolicy": "Always",
              "name": "wait-for-postgres",
              "resources": Object {
                "limits": Object {
                  "cpu": "20m",
                  "memory": "32Mi",
                },
                "requests": Object {
                  "cpu": "5m",
                  "memory": "16Mi",
                },
              },
            },
          ],
          "volumes": Array [
            Object {
              "name": "uploads",
              "persistentVolumeClaim": Object {
                "claimName": "some-pvc",
              },
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "annotations": Object {
        "app.gitlab.com/app": "some-path",
        "app.gitlab.com/env": "some-slug",
        "app.gitlab.com/env.name": "some-name",
        "kapp.k14s.io/disable-default-label-scoping-rules": "",
        "kapp.k14s.io/disable-default-ownership-label-rules": "",
      },
      "labels": Object {
        "app": "hello",
        "application": "some-slug-some-project",
        "owner": "some-project",
        "team": "some-project",
      },
      "name": "hello",
      "namespace": "some-namespace",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "http",
          "port": 80,
          "targetPort": 1337,
        },
      ],
      "selector": Object {
        "app": "hello",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "extensions/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "annotations": Object {
        "app.gitlab.com/app": "some-path",
        "app.gitlab.com/env": "some-slug",
        "app.gitlab.com/env.name": "some-name",
        "kapp.k14s.io/disable-default-label-scoping-rules": "",
        "kapp.k14s.io/disable-default-ownership-label-rules": "",
        "kubernetes.io/ingress.class": "nginx",
      },
      "labels": Object {
        "app": "hello",
        "application": "some-slug-some-project",
        "owner": "some-project",
        "team": "some-project",
      },
      "name": "hello",
      "namespace": "some-namespace",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "strapi-some-slug-some-project.dev2.fabrique.social.gouv.fr",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "hello",
                  "servicePort": 80,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "strapi-some-slug-some-project.dev2.fabrique.social.gouv.fr",
          ],
          "secretName": "wildcard-crt",
        },
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "PersistentVolumeClaim",
    "metadata": Object {
      "annotations": Object {},
      "name": "some-pvc",
    },
    "spec": Object {
      "accessModes": Array [
        "ReadWriteOnce",
      ],
      "resources": Object {
        "requests": Object {
          "storage": "5Gi",
        },
      },
      "volumeMode": "Filesystem",
    },
  },
]
`;
